Add documentation for running scripts (#4968)
astral-sh/uv: docs/guides/scripts.md

# Running scripts

> 

Reviewer: Should we add a bit of top-level context about scripts and the benefit of running them through `uv`?

## Running a script without dependencies

If a script has no dependencies, it can be executed with `uv run`:

```python
print("Hello world")
```

```console
$ uv run example.py
Hello world
```

Similarly, if the script depends on a module in the standard library, there's nothing more to do:

```python
import os

print(os.path.expanduser("~"))
```

```console
$ uv run example.py
/Users/astral
```

Arguments can be passed to the script:

```python
import sys

print(" ".join(sys.argv[1:]))
```

```console
$ uv run example.py test
test

$ uv run example.py hello world!
hello world!
```

Note that if `uv run` is used in a _project_, i.e. a directory with a `pyproject.toml`, it will install the current project before running the script. If the script does not depend on the project, use the `--isolated` flag to skip this:

```console
# Note, it is important that the flag comes _before_ the script
$ uv run --isolated example.py
```

See the [projects](./projects.md) guide for more details on working in projects.

## Running a script with dependencies

> When a script requires dependencies, they must be installed into the environment that the script runs in. uv prefers to create these environments on-demand instead of having a long-lived virtual environment with manually managed dependencies. This requires declaration

Reviewer: ```suggestion
When a script requires dependencies, they must be installed into the environment that the script runs in. uv prefers to create these environments on-demand instead of maintaining a long-lived virtual environment with manually managed dependencies. This requires explicit declaration
```

For example, the following script requires `rich`.

```python
import time
from rich.progress import track

for i in track(range(20), description="For example:"):
    time.sleep(0.05)
```

If executed without specifying a dependency, this script will fail:

```console
$ uv run --isolated example.py
Traceback (most recent call last):
  File "/Users/astral/example.py", line 2, in <module>
    from rich.progress import track
ModuleNotFoundError: No module named 'rich'
```

The dependency can be requested with the `--with` option:

```console
$ uv run --with rich example.py
For example: ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:01
```

Constraints can be added to the requested dependency if specific versions are needed:

```consoleq
$ uv run --with 'rich>12,<13' example.py
```

Multiple dependencies can be requested by repeating with `--with` option.

Note that if `uv run` is used in a _project_, these dependencies will be included _in addition_ to the project's dependencies. To opt-out of this behavior, use the `--isolated` flag.

> ## Declaring script dependencies

Author: Somewhere around here we'd cover #4973

Python recently added a standard format for [inline script metadata](https://packaging.python.org/en/latest/specifications/inline-script-metadata/#inline-script-metadata). This allows the dependencies for a script to be declared in the script itself.

To use inline script metadata, include a `script` section at the top of the script:

```python
# /// script
# dependencies = [
#   "requests<3",
#   "rich",
# ]
# ///

import requests
from rich.pretty import pprint

resp = requests.get("https://peps.python.org/api/peps.json")
data = resp.json()
pprint([(k, v["title"]) for k, v in data.items()][:10])
```

> uv will automatically create an environment with the dependencies when running the script, e.g.:

Reviewer: ```suggestion
uv will automatically create an environment with the dependencies necessary to run the script, e.g.:
```
